### DISCUSSION TITLE
*random* connectivity problems on a customers vm

### DISCUSSION THREAD

**Discussion by leonmeerfeld** [prisma/prisma#18802] [Q&A]
Hello everyone,

I have this weird issue with prisma. On a customers vm (sql server) database connections only are working sometimes.
This problem does not occur on my dev machine or other vms.
The application will be working for a bit, but once I restart oder change something im getting the following error:

Timed out fetching a new connection from the connection pool. More info: ...
(Current connection pool timeout: 10, connection limit 9)

I don't think its an connection pool issue, since there is no open connection for the db user am using (when I check sys.sysprocesses or sys.dm_exec_sessions tables) It seems like prisma has a problem creating a new connection on occasion.
Also I can restart the database or vm and the issue is still not resolved. It does work sometimes after I change some things (not reliably though, only for a short time).

The database connection is stable and fast if its working. Also the db is running on the same machine.
Things like npx prisma db pull also don't work (Can't reach database at ...) although the connection to the database is stable.
Other applications or other mssql nodejs application that don't use prisma do not have this issue.

When it's working its really fast so I don't think its a timeout issue, the vm is also powerful.

If it is a connection pooling issue, how would I go about fixing this?

Thanks in advance and let me know if you need more info that I can provide.

**Comment by nurul3101** [comment]
Hello @leonmeerfeld 👋 

As a first step, I would recommend enabling [metrics](https://www.prisma.io/docs/concepts/components/prisma-client/metrics) as it would provide you with all needed information on the connections opened by prisma.

Can you enable metrics and check if there are any busy or waiting connections?

<img width="695" alt="Screenshot 2023-04-18 at 5 54 38 PM" src="https://user-images.githubusercontent.com/13049676/232776408-f7b58f5e-1d20-4a79-b198-7d608b4909e4.png">

**Reply by leonmeerfeld** [reply]
Thank you for the fast reply. I did that and got this:
This is before the queries fail:

metrics before {
  counters: [
    {
      key: 'prisma_client_queries_total',
      labels: {},
      value: 27,
      description: 'Total number of Prisma Client queries executed'
    },
    {
      key: 'prisma_datasource_queries_total',
      labels: {},
      value: 27,
      description: 'Total number of Datasource Queries executed'
    },
    {
      key: 'prisma_pool_connections_open',
      labels: {},
      value: 1,
      description: 'Number of currently open Pool Connections'
    }
  ],
  gauges: [
    {
      key: 'prisma_client_queries_active',
      labels: {},
      value: 0,
      description: 'Number of currently active Prisma Client queries'
    },
    {
      key: 'prisma_client_queries_wait',
      labels: {},
      value: 0,
      description: 'Number of Prisma Client queries currently waiting for a connection'
    },
    {
      key: 'prisma_pool_connections_busy',
      labels: {},
      value: 0,
      description: 'Number of currently busy Pool Connections (executing a datasource query)'
    },
    {
      key: 'prisma_pool_connections_idle',
      labels: {},
      value: 9,
      description: 'Number of currently unused Pool Connections (waiting for the next datasource query to run)'
    },
    {
      key: 'prisma_pool_connections_opened_total',
      labels: {},
      value: 1,
      description: 'Total number of Pool Connections opened'
    }
  ],
  histograms: [
    {
      key: 'prisma_client_queries_duration_histogram_ms',
      labels: {},
      value: [Object],
      description: 'Histogram of the duration of all executed Prisma Client queries in ms'
    },
    {
      key: 'prisma_client_queries_wait_histogram_ms',
      labels: {},
      value: [Object],
      description: 'Histogram of the wait time of all Prisma Client Queries in ms'
    },
    {
      key: 'prisma_datasource_queries_duration_histogram_ms',
      labels: {},
      value: [Object],
      description: 'Histogram of the duration of all executed Datasource Queries in ms'
    }
  ]
}


I think it might be a pooling issue. I also noticed the following:
Initially I load a bunch of data from different tables. Its atleast 11 queries. If I reduce it to 9 or less queries, it works. After that its not working anymore ("Cannot read properties of undefined (reading 'findMany')".
But this only occurs on the "normal" sql server instance. I've also installed sql express for testing purposes and I don't have the same problem there.

**Reply by nurul3101** [reply]
Can you try increasing the [connection_limit](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#increasing-the-pool-size) and check if you get the same error?

**Reply by nurul3101** [reply]
@leonmeerfeld If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏
